crates/test_probe/src/main.rs

```diff
@@ -30,13 +85,29 @@ fn main() -> Result<()> {
     let speedup = slowdown.clone().rev();
 
     for delay_ms in speedup.chain(slowdown).cycle() {
-        if let Err(err) = client.send(&packet) {
-            eprintln!("Could not send event {:?}", err)
-        };
-        thread::sleep(Duration::from_millis(delay_ms.into()));
+        send(&packet);
+        sleep(Duration::from_millis(delay_ms.into()));
+    }
+
+    unreachable!()
+}
+
+fn burst(burst_period: Duration, events_per_burst: u32, send: impl Fn(&Packet)) -> ! {
+    let event_period = burst_period / events_per_burst;
+    // Delay events' timestamps by one burst period to ensure they're not set in future when sent out.
+    let event_start = current_timestamp() - burst_period;
+    let mut event_generator =
+        (0..).map(|n| Event::with_timestamp(EventKind::TestTick, event_start + n * event_period));
+
+    // Prevent drifting away from the given frequency by computing the sleep duration for each cycle.
+    let start = Instant::now();
+    for deadline in (0..).map(|i| start + i * burst_period) {
```

Comment:
We use this deadline-based looping on 3 places already. 

Maybe that's about time to pull it out to a shared utility in composer-api's util module?

Reply:
haha, and I see now that I even repeated the same bug here. 🙈 (Fixed by a force-push.)

I guess that's another reason for pulling this out. .)